src/main/Main.hs

```diff
@@ -470,21 +473,41 @@ setupCmd SetupCmdOpts{..} go@GlobalOpts{..} = do
               )
           Nothing
 
-withConfig :: GlobalOpts
+-- | Enforce mutual exclusion of every action running via this function
+-- on this users account.  Currently, stack uses this to completely
+-- exclude concurrent execution of stack commands.  In the future,
+-- stack may refine this to a finer-grained locking approach.
+withUserFileLock :: Config
+                 -> (FileLock -> IO a)
+                 -> IO a
+withUserFileLock cfg act = do
+    lockfile <- parseRelFile "lockfile"
```

Comment:
Typically we use `$(mkRelFile "lockfile")` for these kinds of things to avoid a runtime check and catch typos at compile time as well.